### PR TITLE
CLI: a splash of colour

### PR DESCRIPTION
**Pull request by oliver-sanders**
#### Add colour support to the `cli_function` wrapper.

* Use the GNU-like colour argument for all Cylc sub-commands:
  * i.e. --color=(auto|always|never)
  * Default to `auto` which will use colour if displayed in a terminal.
* Use the `ansimarkup` library for easy colour control:
  * An xml-like pre-processor with basic ANSI style and 8-bit coluor support.
  * e.g. `<red>Error</red>`.
* Make important information prominent and unimportant information ambient.
   * CRITICAL - bold & red
   * ERROR - red
   * WARNING - yellow
   * DEBUG - grey
   * Success messages - green
   * Everything else - black
* Improve logger output when viewed from the terminal (log/suite/log file not effected).
  * Log entries formatted to width of terminal.
  * Critical, error and warning levels coloured.
* Colorise log/suite/log when viewed through `cylc cat-log`
* Convert some `sys.exit` to `raise CylcError`

*Note: The `cli_function` wrapper may change with #2972, colour support is somewhat orthogonal to this.*

#### Example

Screenshot from 2019-06-20 13-30-11

### CODE REVIEW

**Comment by hjoliver** [comment]
Some general comments first:

I definitely like colour output for the CLI 

Colorama docs say: "ANSI escape character sequences have long been used to produce colored terminal text and cursor positioning on Unix and Macs. Colorama makes this work on Windows, too, by wrapping stdout, stripping ANSI sequences it finds (which would appear as gobbledygook in the output), and converting them into the appropriate win32 calls to modify the state of the terminal. On other platforms, Colorama does nothing."

i.e. "[on Linux] Colorama does nothing", and its main purpose is to translate ANSI color control codes to the MS Windows API (which we don't support).

Presumably in spite of the "it does nothing [on Linux]" statement, it is nicer to use than raw ANSI control codes.

> There are high level libs built ontop of colorama including my fave ansimarkup.

Should we use ansimarkup then??

**Comment by oliver-sanders** [comment]
No objections so I've moved this along:

* Using [ansimarkup](https://github.com/gvalkov/python-ansimarkup) to simplify output e.g. `<red>red text</red>`
* Pass the option parser to `cli_function` which populates the arguments of its decorated function

Colour support so far:

* `cylc broadcast`
* `cylc validate`
* `cylc show`

If no-one screams I'll complete the conversion, rename `cli_function2` to `cli_function` and open to review.

@kinow I'm guessing you're not going to like my wrapping of the CLI methods as the wrapper populates the arguments. Happy to change, open to suggestions, trying not to write any solution into `argparse`.

**Comment by kinow** [comment]
Tested with a running suite, then validate and show, and show got some colours (validate was green, same colour as my terminal background, so maybe there was some colouring, but I couldn't notice).

>Using ansimarkup to simplify output e.g. <red>red text</red>

Does it add/help a lot? Last commit in 2018, and few contributors - which could be fine, as the tool is simple, it could be feature-complete and only fix bugs (which it doesn't have any at the moment in github). But if it does add or simplify the code, and you think it's stable to be used in Cylc, all good :+1: 

>@kinow I'm guessing you're not going to like my wrapping of the CLI methods as the wrapper populates the arguments. Happy to change, open to suggestions, trying not to write any solution into argparse.

No suggestion. I have never used anything like this before, so it looks quite alien for me, but not incorrect. I think it is OK to be added now and we can think in alternatives later if we find issues with the current approach, or if there are simpler alternatives.

**Comment by oliver-sanders** [comment]
> validate was green, same colour as my terminal background

You have a green terminal background!

Not too sure what can be done about this one, it might be possible to detect the terminal background ANSI code, then pair the corresponding foreground code with a white background.

What do you see if you see with tools like `diff`?

> ansimarkup - Does it add/help a lot?

It simplifies things somewhat making code more readable (cleaner syntax, no string formatting required), it also adds 8-bit colour support (where the terminal supports it) as used to turn the DEBUG messages grey (there is no grey ANSI code).

I think `ansimarkup` suffers from "being too simple" so it's not active as there aren't really any important changes to make.

> looks quite alien for me, but not incorrect. I think it is OK

I'm not overjoyed by it as it causes linters to complain, but it's a perfectly valid Python pattern. Happy to change later for a nicer approach if we come up with one.

**Comment by kinow** [comment]
Oh sorry, actually green text color! Hahaa, would be pretty hard on my eyes a green background haha

**Comment by oliver-sanders** [comment]
Flaky `tests/execution-time-limit/00-background.t` keeps failing but all else passes.

File: bin/cylc-cat-log [hunk 1]
```diff
@@ -109,6 +109,34 @@ MODES = {
...
 
 
+def colorise_cat_log(cat_proc, color=False):
+    """Print a Cylc log file in color at it would appear in the terminal.
+
+    Args:
+        cat_proc (subprocess.Process):
+            A process which outputs a Cylc log file i.e. `cat`.
+        color (bool):
+            If `True` log will appear in color, if `False` no controll
```
Comment (hjoliver) [inline bin/cylc-cat-log]: I believe "controll" is spelled "controlll".

**Comment by hjoliver** [comment]
Codacy failure (2 new issues) because Pylint evidently doesn't understand when decorators alter a function signature:

Screenshot from 2019-06-25 16-58-47

Which is annoying.

File: cylc/flow/scheduler_cli.py [hunk 2]
```diff
@@ -231,10 +177,83 @@ def parse_commandline(is_restart):
...
+        scheduler = Scheduler(is_restart, options, args)
+    except SuiteServiceFileError as exc:
+        sys.exit(exc)
+    scheduler.start()
+
+
+def main(is_restart=False):
+    """Abstraction for cylc (run|restart) CLI"""
+    if is_restart:
+        return restart()
```
[the remark this answers is not in the document]
Comment (hjoliver) [reply cylc/flow/scheduler_cli.py]: (Plus an explanatory comment: "Pylint bugs doesn't borks when function signature changed by decorator" or something like that).

**Comment by hjoliver** [comment]
Otherwise LGTM (not that the codacy problem is your fault; Pylint is just deficient).

**Comment by hjoliver** [comment]
Spurious test battery failure on Travis CI.

**Comment by oliver-sanders** [comment]
:rainbow:

**Pull request merged**